lib/literal/enum.rb

```diff
@@ -104,7 +149,24 @@ def name
 
 	alias_method :inspect, :name
 
+	def deconstruct
+		[value]
+	end
+
+	def deconstruct_keys(keys)
+		h = to_h
+		keys ? h.slice(*keys) : h
+	end
+
 	def _dump(level)
 		Marshal.dump(@value)
 	end
 end
+
+TracePoint.trace(:end) do |tp|
```

Comment:
FWIW, it seems very inefficient to use a TracePoint for this, because it gets called for every `class/module ... end`.
The `inherited` hook seems much better for this as then it's only called for subclasses of Literal::Enum.

Reply:
Ah but if it's used for freezing the class then `inherited` will be called too early I suppose.
Maybe the freezing could be done via Class#subclasses instead or so then, or an explicit call at the end of defining a Literal::Enum.

Reply:
Unfortunately this does need to happen after the class is fully defined. The `inherited` hook is called too early.

I’m not sure how `Class#subclasses` would work here. What would be the trigger?

I’ve benchmarked this in CRuby and it doesn't seem to make much difference since `end` is only triggered a for each class and most apps don’t define that many classes. It’s usually a few thousand, not hundreds of millions and it all happens at boot.

One potential optimisation would be to create the tracepoint on `inherited` and then disable it after it is hit with the subclass it’s looking for. I believe a TracePoint can disable itself.

Reply:
> I’m not sure how `Class#subclasses` would work here. What would be the trigger?

Something like `Process.warmup` maybe, but it's not ideal.

Maybe this initialization could be done on the first usage of that enum lazily?
Then you wouldn't need any hook.

> I’ve benchmarked this in CRuby and it doesn't seem to make much difference since `end` is only triggered once for each class defined, and most apps don’t define that many classes. It’s usually a few thousand, not hundreds of millions and they all happen at boot.

It used to be the case that every TracePoint would disable the JIT or throw away JITed code.
`:class` and `:end` are only called from "slow/class definition" paths so they don't really need to do that, but then they also need to be handled specially and not like other TracePoints internally.
I would think creating a Binding per class and calling the TracePoint is a non-trivial overhead when an application has thousands of classes.

> One potential optimisation would be to create the tracepoint on `inherited` and then disable it after it is hit with the subclass it’s looking for. I believe a TracePoint can disable itself.

Depending on the implementation of TracePoints, enabling/disabling can be very expensive too and cause deoptimization.

---

I was also thinking maybe it could be something like:
```ruby
Color = Literal::Enum(Integer) do
  Red = new(1)
  Green = new(2)
  Blue = new(3)
end
```
but that would assign the constants to the top-level scope, so it'd need to be:
```ruby
Color = Literal::Enum(Integer) do
  self::Red = new(1)
  self::Green = new(2)
  self::Blue = new(3)
end
```
or
```ruby
Color = Literal::Enum(Integer) do
  value :Red, 1
  value :Green, 2
  value :Blue, 3
end
```
That would then naturally give you a hook after the Enum is done being defined.
(and it would also not call this hook twice if the enum class is opened in multiple places)

Reply:
The more dynamic we get here, the less we are able to support language servers. It’s important that language servers pick up these constants and know what type they are.

Reply:
@eregon: If you are worried about performance in a certain kind of scenario, I would recommend writing a reproducible benchmark script that demonstrates a performance issue. This will (1) prove a bottleneck exists and (2) provide a check that a "fix" indeed fixes the issue. 

Otherwise, we are talking about theoretical concerns with theoretical possible improvements.

Reply:
We could do without the class freezing, but we would still need a trigger to generate the indexes. We never know when the final mutation occurs so we would have to generate indexes on every mutation. If you have an enum with a few hundred members, that means generating indexes a few hundred times, which is probably more overhead than waiting for the end trace point.

Also, unless I've missed something, YJIT doesn't optimise boot time. It's not even recommended to enable YJIT until after boot.

Reply:
I benchmarked 1,000 class definitions with and without the tracepoint in CRuby. In the case without the tracepoint, I manually called `__after_defined__` at the end of the class. There’s ~34µs overhead per enum class defined.

With TracePoint: 145ms
Without TracePoint: 111ms